llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -1718,6 +1718,25 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
         Builder.CreateIntrinsic(Intrinsic::umax, {I.getType()}, {A, B}));
   }
 
+  // ctpop(A | B) + ctpop(A & B) -> ctpop(A) + ctpop(B)
+  if ((match(LHS, m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
+                      m_And(m_Value(A), m_Value(B))))) &&
+       (match(RHS, m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
+                       m_Or(m_Specific(A), m_Specific(B))))) ||
+        match(RHS, m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
+                       m_Or(m_Specific(B), m_Specific(A))))))) ||
+      (match(LHS, m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
+                      m_Or(m_Value(A), m_Value(B))))) &&
+       (match(RHS, m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
+                       m_And(m_Specific(A), m_Specific(B))))) ||
+        match(RHS, m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
+                       m_And(m_Specific(B), m_Specific(A)))))))) {
```

Comment:
```suggestion
  if (match(&I, m_c_Add(m_OneUse(m_Intrinsic<Intrinsic::ctpop>(
                       m_And(m_Value(A), m_Value(B))))), mOneUse(m_Intrinsic<Intrinsic::ctpop>(
                       m_c_Or(m_Specific(A), m_Specific(B)))))))) {
```

Reply:
Tried this and now the transforms don't work anymore

Reply:
> Tried this and now the transforms don't work anymore

Oh, you need `m_Deferred` instead of `m_Specific`. Sorry for my mistake.